### DISCUSSION TITLE
Apparent inability to concurrently write to files

### DISCUSSION THREAD

**Discussion by AGogosha** [vlang/v#17254] [Questions and Answers] [answered]
Welcome back to another installment of a physics major attempting to learn to program, get ready for a long one again. My question is essentially the title of this post, what is the intended way to concurrently call functions that write out to a file? This question comes from my attempt to rewrite the code I previously asked a question about [here](https://github.com/vlang/v/discussions/17230) by calculating the simulation path and writing the coordinates out to a file concurrently instead of saving the path in an array and then saving the path to a file later. Now my code for this can be found [here](https://github.com/AGogosha/ball_simulation/tree/master/src) under src/main_v2.v.
Now onto the discussion of my question. In the example code for how to manage concurrent threads found [here in the documentation](https://github.com/vlang/v/blob/master/doc/docs.md#concurrency) you push the threads to an array of threads with the same out type as the function that is being put into the array. This becomes an issue when the [example code for how to use a defer statement](https://github.com/vlang/v/blob/master/doc/docs.md#defer), the only mention in the documentation of writing text to a file that I can find, returns a !int which is not a type that can be passed to an array. This is also an issue in my code that I linked about that does something similar. This results in it appearing that there is no way to concurrently write to a file.
Now there are a couple of ways of thinking about the appearance of this issue. That you really can do that and I'm just dumb and can't find a way to do this. Two, that you can do that and the documentation just doesn't explain how to do that. Or 3, you really can't concurrently write to multiple files. If this is the case then one immediate solution is to not have writeln and other similar functions not return anything. If there is a good reason for these functions to return anything please tell me so I can stop thinking that this is really dumb. The other solution is to allow arrays to hold types like ! and ?. This last point comes from my heavy use of python and how they work with files.
Thank you all for coming to my ted talk, I hope you understand what my issue is and don't just think I'm crazy.

**Comment by MatejMagat305** [comment]
Only crazy idea: try `-gc none -autofree` and write what happen
because GC litle bit limit thread to ~1.2 core, I don't know why, but ...

**Comment by j-bm** [comment, marked as answer]
_...there is no way to concurrently write to a file..._

V (and many other languages) rely on the operating system features and functions to implement many details such as this.  In general, file operations are single-threaded and this is a constraint of the operating system, largely because it has to maintain buffers and data streams in progress (pointers to the last byte written, for example.)  And this is very hard, so operating system authors decline to implement this.

Getting past operating system constraints such as this is a vast topic.  Threads (and go-routines and so on) are one example where a single-threaded process of an operating system was considered insufficient so Linux and others added various kinds of threads and other forms of concurrency.  This happened over the course of the last 10 or 15 years or so.

File IO however needs a different trick: write your results not to a file but to a channel and create one thread dedicated to copying results from the channel to the file.  To distinguish results computed by different threads include the thread number in the data on the channel.

This adapts your code to a single-threaded writer (which is what the operating system likes to see) fed by many concurrent threads (which is what your program likes to do.)
